### ISSUE TITLE
[R] Windows builds are failing at the link stage

### ISSUE THREAD

**Issue by paleolimbot** [apache/arrow#36707]
### Describe the bug, including details regarding any error messages, version, and platform.

Now happening on main, e.g., https://github.com/apache/arrow/actions/runs/5566743012/jobs/10168304989?pr=36706#step:12:91

```
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libarrow_bundled_dependencies.a(hash_function_impl.cc.obj):(.text+0x26c1): undefined reference to `EVP_MD_size'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libcurl.a(libcurl_la-openssl.o):(.text+0x2af2): undefined reference to `EVP_PKEY_id'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libcurl.a(libcurl_la-openssl.o):(.text+0x53b5): undefined reference to `SSL_get_peer_certificate'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libcurl.a(libcurl_la-openssl.o):(.text+0x5e0a): undefined reference to `SSL_get_peer_certificate'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libcurl.a(libcurl_la-openssl.o):(.text+0x5fe9): undefined reference to `EVP_PKEY_id'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libssh2.a(crypto.o):(.text+0x3e62): undefined reference to `EVP_PKEY_id'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libssh2.a(crypto.o):(.text+0x5676): undefined reference to `EVP_PKEY_id'
C:\rtools42\x86_64-w64-mingw32.static.posix\bin/ld.exe: ../windows/arrow-12.0.1.9000/lib/x64-ucrt/libssh2.a(crypto.o):(.text+0x5b50): undefined reference to `EVP_PKEY_id'
```

This feels like an OpenSSL linking issue but I'm not (yet) familiar with the details of how this all works on Windows.

### Component(s)

R

**Comment by kou** [comment]
This is also caused by mixing OpenSSL:

OpenSSL 3 is provided by R's MSYS2 packages recently: https://github.com/apache/arrow/actions/runs/5566743012/jobs/10168091288?pr=36706#step:7:215

```text
downloading mingw-w64-ucrt-x86_64-openssl-3.1.1-9800-any.pkg.tar.xz...
```

But OpenSSL 1 provided by GitHub Actions runner is used: https://github.com/apache/arrow/actions/runs/5566743012/jobs/10168091288?pr=36706#step:7:454

```text
-- Found OpenSSL: C:/Program Files/OpenSSL/lib/libcrypto.lib (found suitable version "1.1.1u", minimum required is "1.0.2")  
```

**Comment by raulcd** [comment]
Should I wait for this to be merged for 13.0.0?
